### PR TITLE
Add Scroll integration

### PR DESCRIPTION
**Pull request by makoto**
## List of changes

- Change `StateProof` interface from `  stateTrieWitness: string[];,  storageProofs: string[][];` to `  stateTrieWitness: string;,  storageProofs: string[];`
- Modify `getStorageValues` to take a proof verification function which isn't dependent on SecureMerkleTrie
- Move `getStorageRoot` from `EVMHelper.sol` to `MerkleTrieProofHelper.sol`

## Gas comparison

Arbitrum

```
  ArbVerifier
Gas estimate 770439
    ✔ simple proofs for fixed values (2154ms)
Gas estimate 722247
    ✔ simple proofs for dynamic values (1314ms)
Gas estimate 777244
    ✔ nested proofs for dynamic values (1361ms)
Gas estimate 1045721
    ✔ nested proofs for long dynamic values (1437ms)
Gas estimate 887013
    ✔ nested proofs with lookbehind (1206ms)
Gas estimate 940514
    ✔ nested proofs with lookbehind for dynamic values (1631ms)
Gas estimate 726250
    ✔ mappings with variable-length keys (1337ms)
Gas estimate 840324
    ✔ nested proofs of mappings with variable-length keys (1479ms)
Gas estimate 679429
    ✔ treats uninitialized storage elements as zeroes (1040ms)
Gas estimate 683462
    ✔ treats uninitialized dynamic values as empty strings (1426ms)
```

Scroll
```
  ScrollVerifier
Gas estimate 1187799
    ✔ simple proofs for fixed values (2016ms)
Gas estimate 1188778
    ✔ simple proofs for dynamic values (1545ms)
Gas estimate 1128781
    ✔ nested proofs for dynamic values (1849ms)
Gas estimate 3391568
    ✔ nested proofs for long dynamic values (1673ms)
Gas estimate 2294214
    ✔ nested proofs with lookbehind (1449ms)
Gas estimate 2231340
    ✔ nested proofs with lookbehind for dynamic values (1838ms)
Gas estimate 1160883
    ✔ mappings with variable-length keys (1364ms)
Gas estimate 2295970
    ✔ nested proofs of mappings with variable-length keys (1510ms)
Gas estimate 1093162
    ✔ treats uninitialized storage elements as zeroes (1357ms)
Gas estimate 1097201
    ✔ treats uninitialized dynamic values as empty strings (1488ms)
```

## Things to improve

- `ScrollProofService` is currently fetching the `batch_index` via centralised API server. Should replace it with l1 smart contract call once Scroll team provides the functionality.
- `verifier.verifyZkTrieProof` currently [verifies account proof every time it tries to verify storage proof](https://github.com/scroll-tech/scroll/blob/612e66ebc201cb504ee0f518ed91405d88eec944/contracts/src/libraries/verifier/ZkTrieVerifier.sol#L259) which is unnecessary. We are currently requesting Scroll team to provide function that can only verify account proof or storage proof, not both at the same time.

### CODE REVIEW

File: evm-gateway/src/utils.ts [hunk 1]
```diff
@@ -22,3 +24,22 @@ export const propsDecoder: PropsDecoder<CFWRequest> = (
...
+    const storageProof = AbiCoder.defaultAbiCoder().encode(
+      ['bytes[]'],
+      [proof.storageProofs[index]]
+    );
+    storageProofs[index] = storageProof;
+  }
+  return {
+    stateTrieWitness,
+    storageProofs,
+  };
```
Comment (Arachnid) [inline evm-gateway/src/utils.ts:44]:
```suggestion
  def flatten(data: string[]) {
    return AbiCoder.defaultAbiCoder().encode(['bytes[]'], [data]);
  }
  return {
    stateTrieWitness: flatten(proof.stateTrieWitness),
    storageProofs: proof.storageProofs.map(flatten),
  };
```

File: evm-verifier/contracts/MerkleTrieProofHelper.sol [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * @title MerkleTrieProofHelper
+ * @notice MerkleTrieProofHelper is a helper library that has functions to interact
+ *         with SecureMerkleTrie.
+ */
+library MerkleTrieProofHelper {
+    error AccountNotFound(address);
+
+    /*
+     * @notice Get the storage value for the provided merkle proof
+     * @param address Not in use. Only exist to provide consistent function argument
```
Comment (Arachnid) [inline evm-verifier/contracts/MerkleTrieProofHelper.sol:18]:
```suggestion
     * @param address Unused. Required so the function signature matches the one required by `EVMProofHelper`
```

File: scroll-gateway/src/ScrollProofService.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+     * @param slots An array of slots to fetch data for.
+     * @returns A proof of the given slots, encoded in a manner that this service's
+     *   corresponding decoding library will understand.
+     */
+    async getProofs(
+        block: ScrollProvableBlock,
+        address: AddressLike,
+        slots: bigint[]
+    ): Promise<string> {
+        const searchUrl = 'https://sepolia-api-re.scroll.io/api/search';
```
Comment (Arachnid) [inline scroll-gateway/src/ScrollProofService.ts:43]: This definitely needs to be a constructor argument, and ultimately supplied via the command line.

File: scroll-gateway/src/ScrollProofService.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                    storageProofs:compressedProofs
+                },
+            ]
+        );
+    }
+  /**
+   * @dev Returns an object representing a block whose state can be proven on L1.
+   */
+  public async getProvableBlock(): Promise<ScrollProvableBlock> {
+        const block = await this.l2Provider.send("eth_getBlockByNumber", ["finalized", false]);
```
Comment (Arachnid) [inline scroll-gateway/src/ScrollProofService.ts:78]: Shouldn't `finalized` be `true`?

File: scroll-verifier/contracts/ScrollVerifier.sol [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+     * @returns {string[]} An array containing the gateway URLs.
+     *     */
+    function gatewayURLs() external view returns (string[] memory) {
+        return _gatewayURLs;
+    }
+
+
+    function getTrieProof(address target, uint256 slot, bytes memory compressedProof, bytes32 root) internal view returns(bytes memory){
+        (bytes32 stateRoot, bytes32 storageValue) = verifier.verifyZkTrieProof(target, bytes32(slot), compressedProof);
+        require(stateRoot == root, "stateRoot not matched");
```
Comment (Arachnid) [inline scroll-verifier/contracts/ScrollVerifier.sol:60]: This should throw an error object.

File: scroll-verifier/contracts/ScrollVerifier.sol [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+     * @param {bytes} proof - The proof data containing Scroll witness data and state proof.
+     */
+    function getStorageValues(
+        address target,
+        bytes32[] memory commands,
+        bytes[] memory constants,
+        bytes memory proof
+    ) external view returns (bytes[] memory values) {
+        (ScrollWitnessData memory scrollData, StateProof memory stateProof) = abi.decode(proof, (ScrollWitnessData, StateProof));
+        bytes32 expectedStateRoot = IScrollChain(verifier.rollup()).finalizedStateRoots(scrollData.batchIndex);
```
Comment (Arachnid) [inline scroll-verifier/contracts/ScrollVerifier.sol:79]: We need to add support for checking the block is the right age.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: evm-gateway/src/utils.ts (hunk 1 above)
Comment (makoto) [reply evm-gateway/src/utils.ts]: Did you mean `function` instead of `def`?

File: scroll-gateway/src/ScrollProofService.ts (hunk 4 above)
Comment (makoto) [reply scroll-gateway/src/ScrollProofService.ts:80]: I don't think it means `finalized` to be `false`

According to the doc (https://www.ankr.com/docs/rpc-service/chains/chains-api/scroll/#eth_getblockbynumber) ,

```
<boolean>: if true it returns the full transaction objects, 
if false — only the hashes of the transactions.
```

both L1ProofService and OPProofService set the second value as `false` as well FYI

File: scroll-verifier/contracts/ScrollVerifier.sol (hunk 6 above)
Comment (makoto) [reply scroll-verifier/contracts/ScrollVerifier.sol:82]: How did you handle on L1 and OP Verifier? I didn't find any right age check on these verifiers.

File: scroll-verifier/deploy_l1/00_scroll_verifier.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+const GATEWAY_URLS = {
+  scrollDevnetL1: 'http://localhost:8089/{sender}/{data}.json',
+  goerli: 'https://scroll-gateway-worker.ens-cf.workers.dev/{sender}/{data}.json',
+  // Point to localhost if you want to test locally
+  // sepolia: 'http://localhost:8080/{sender}/{data}.json'
+  sepolia: 'https://scroll-sepolia-gateway-worker.ens-cf.workers.dev/{sender}/{data}.json',
+};
+
+const ROLLUP_ADDRESSES = {
+  sepolia: '0x64cb3A0Dcf43Ae0EE35C1C15edDF5F46D48Fa570',
```
Comment (zimpha) [inline scroll-verifier/deploy_l1/00_scroll_verifier.ts:14]: This should be the final version used in sepolia: https://sepolia.etherscan.io/address/0xE0BfA7f3B06A9589A914BE09Ba0E5671f481A722#code.

File: scroll-verifier/contracts/ScrollVerifier.sol (hunk 6 above)
Comment (makoto) [reply scroll-verifier/contracts/ScrollVerifier.sol:82]: This is tracked separately at https://github.com/ensdomains/evmgateway/issues/47

**Pull request merged**